// @semcom/command-version/index.js
import { isolate } from 'transmutation'
import setupGit from '@kikd/ci-setup-git'
import deployGit from '@kikd/ci-deploy-git'
import getPackages from '@kikd/command-list'
import getChanges from '@semcom/command-changes'
import getDependencies from '@kikd/command-deps'
import updatePackages from '@semcom/update-packages'
import getPackageDetails from '@kikd/package-details'
import template from '@semcom/template-update-packages'
import recommendVersions from '@semcom/recommend-version'

const { CI } = process.env

export default function version(config) {
    return getPackages({ ...config, changedAndPublishable: true })
        .extend(getChanges(config).then(isolate(['changes', 'latest'])))
        .extend(
            'dependencies',
            getDependencies(config).then(isolate('dependencies'))
        )
        .extend('manifests', getPackageDetails)
        .extend('updates', recommendVersions)
}

export const command = () => ({
    name: 'version',
    aliases: ['update'],
    run: version,
    template,
    effect: () => config => {
        if (!CI)
            throw new Error('Committing new versions should only happen in CI.')

        return setupGit()
            .then(() => updatePackages(config))
            .then(() => deployGit())
    },
})
