// @kikd/command-init/index.js
/* eslint-disable no-console */
import { existsSync } from 'fs'
import { join, basename } from 'path'

import transmute from 'transmutation'
import template from '@kikd/template-initialize'

const isGitDirCheck = dir => existsSync(join(dir, '.git'))
const packageJsonExistsCheck = dir => existsSync(join(dir, 'package.json'))
const modulesDirExistsCheck = dir => existsSync(join(dir, 'src/node_modules'))

function collectInfo() {
    return transmute({})
        .extend('root', process.cwd())
        .extend('monorepoName', ({ root }) => basename(root))
        .extend('isGitDir', ({ root }) => isGitDirCheck(root))
        .extend('packageJsonExists', ({ root }) => packageJsonExistsCheck(root))
        .extend('modulesDirExists', ({ root }) => modulesDirExistsCheck(root))
}

export default function initialize() {
    // eslint-disable-next-line complexity
    return ({
        monorepoName,
        isGitDir,
        packageJsonExists,
        modulesDirExists,
    }) => {
        if (isGitDir && packageJsonExists && modulesDirExists)
            throw new Error(`[${monorepoName}] is already a KIKD monorepo.`)

        if (!isGitDir) console.log('Initialize Git')
        if (!packageJsonExists) console.log('Create package.json')
        if (!modulesDirExists) console.log('Create src/node_modules')
    }
}

export const command = () => ({
    name: 'initialize',
    run: collectInfo,
    effect: initialize,
    template,
    autoCommit: true,
})
