// @kikd/command-init/index.js
import { join, basename } from 'path'

import git from '@kikd/git'
import log from '@dev-xp/log'
import { write } from '@dev-xp/json'
import transmute from 'transmutation'
import { exists, mkdir } from '@dev-xp/fs'
import template from '@kikd/template-initialize'

const isGitDirCheck = dir => exists(join(dir, '.git'))
const packageJsonExistsCheck = dir => exists(join(dir, 'package.json'))
const modulesDirExistsCheck = dir => exists(join(dir, 'src/node_modules'))

function collectInfo() {
    return transmute({})
        .extend('root', process.cwd())
        .extend('monorepoName', ({ root }) => basename(root))
        .extend('isGitDir', ({ root }) => isGitDirCheck(root))
        .extend('packageJsonExists', ({ root }) => packageJsonExistsCheck(root))
        .extend('modulesDirExists', ({ root }) => modulesDirExistsCheck(root))
}

export default function initialize() {
    // eslint-disable-next-line complexity
    return ({
        root,
        monorepoName,
        isGitDir,
        packageJsonExists,
        modulesDirExists,
    }) => {
        if (isGitDir && packageJsonExists && modulesDirExists)
            throw new Error(`[${monorepoName}] is already a KIKD monorepo.`)

        if (!isGitDir) git.init()

        const manifestPath = join(root, 'package.json')
        if (!packageJsonExists) {
            write(manifestPath, { name: monorepoName })
            log('Package.json created.')
        }

        const modulesPath = join(root, 'src/node_modules')
        if (!modulesDirExists) {
            mkdir(modulesPath)
            log('src/node_modules created.')
        }
    }
}

export const command = () => ({
    name: 'initialize',
    run: collectInfo,
    effect: initialize,
    template,
    autoCommit: true,
})
