// @kikd/command-new/index.js
/* eslint-disable no-console */
import transmute from 'transmutation'
import findRepoRoot from '@kikd/find-repo-root'
import requestPackageType from '@kikd/request-package-type'
import requestPackageName from '@kikd/request-package-name'

function collectInfo(config = {}) {
    const { plugins = {} } = config

    return transmute({})
        .extend('root', findRepoRoot)
        .extend('type', requestPackageType(config))
        .extend('name', requestPackageName)
        .extend('extra', data => {
            const { type } = data
            const plugin = plugins[type]
            if ('preCreate' in plugin) return plugin.preCreate(data)

            return {}
        })
}

export default function create() {
    // eslint-disable-next-line complexity
    return ({ root, type, name }) =>
        console.log(
            `Creating [${name}] package of type [${type}] in ${root}/src/node_modules`
        )
}

export const command = () => ({
    name: 'new',
    run: collectInfo,
    effect: create,
    autoCommit: true,
})
