// @rampant-giveaway/selection/index.js
import { filter, tap, map, withLatestFrom, mergeMap } from 'rxjs/operators'
import weighted from 'weighted'
import omitBy from 'lodash.omitby'
import mapValues from 'lodash.mapvalues'
import { firestore } from 'firebase-admin'
import createCommandStream from '@rampant-giveaway/commands'

const { FieldValue } = firestore

export default ({ moderator, db, prefix, logger: baseLogger }) => {
    const logger = baseLogger.child({ plugin: 'selection' })
    const command$ = createCommandStream(moderator)

    const currentGiveaway$ = db.onDocumentChange('giveaways/current')
    command$
        .pipe(
            filter(
                ({ user_name, command }) =>
                    command === 'give' &&
                    user_name.toLowerCase() === 'raybenefield'
            )
        )
        // eslint-disable-next-line complexity
        .subscribe(({ args }) => {
            let giveaway = null
            let entryPool = prefix
            const slots = 1

            if (args.length === 0) {
                console.log('No arguments given for giveaway.')
            }

            if (args.length === 1) {
                giveaway = args[0]
            }
            if (args.length === 2) {
                giveaway = args[0]
                entryPool = args[1]
            }
            db.set('giveaways/current', { entryPool, giveaway })
            logger.extend(
                `Starting a giveaway with ${entryPool}-entries for ${giveaway} giveaway with ${slots} winner${
                    slots > 1 ? 's' : ''
                }`
            )
        })

    command$
        .pipe(
            filter(({ command }) => command === 'enter'),
            map(({ user_id: userId, user_name: userName }) => ({
                userId,
                userName,
            })),
            withLatestFrom(currentGiveaway$),
            filter(([, { giveaway }]) => giveaway)
        )
        .subscribe(([{ userId, userName }, { entrants = [] }]) => {
            if (entrants.includes(userId)) {
                moderator.whisper(userName, 'You are already in the giveaway.')
                return
            }

            logger.extend(`${userName} entered the giveaway.`)
            moderator.whisper(
                userName,
                'You have been entered into the giveaway.'
            )
            db.update('giveaways/current', {
                entrants: FieldValue.arrayUnion(userId),
            })
        })

    command$
        .pipe(
            filter(
                ({ command, user_name }) =>
                    command === 'pick' &&
                    user_name.toLowerCase() === 'raybenefield'
            ),
            withLatestFrom(currentGiveaway$),
            mergeMap(([, { entryPool, giveaway, entrants }]) =>
                Promise.all([
                    db
                        .getAll(
                            entrants.map(
                                id => `pools/${entryPool}/entries/${id}`
                            )
                        )
                        .then(allEntries =>
                            mapValues(allEntries, ({ name, entries }, id) => ({
                                id,
                                name,
                                entries,
                            }))
                        ),
                    db.get(
                        `giveaways/${entryPool}/giveaways/${giveaway}/winners`
                    ),
                    Promise.resolve(entryPool),
                    Promise.resolve(giveaway),
                ])
            ),
            map(([allEntries, winners, entryPool, giveaway]) => ({
                entryPool,
                giveaway,
                entrants: omitBy(allEntries, ({ id }) => id in winners),
                entries: mapValues(
                    omitBy(allEntries, ({ id }) => id in winners),
                    ({ entries }) => entries
                ),
            })),
            filter(({ entries }) => Object.values(entries).length > 0),
            mergeMap(({ entrants, entries, entryPool, giveaway }) =>
                db
                    .push(
                        `giveaways/${entryPool}/giveaways/${giveaway}/entrants`,
                        entrants
                    )
                    .then(entriesLocation => ({
                        entrants,
                        entries,
                        entryPool,
                        giveaway,
                        entriesLocation,
                    }))
            ),
            map(
                ({
                    entrants,
                    entries,
                    entryPool,
                    giveaway,
                    entriesLocation,
                }) => {
                    const winner = weighted.select(entries)
                    return {
                        winner: {
                            ...entrants[winner],
                            entriesLocation,
                            candidates: Object.keys(entrants).length,
                            allEntries: Object.values(entrants).reduce(
                                (all, { entries: currentEntries = 0 }) =>
                                    currentEntries + all,
                                0
                            ),
                        },
                        entryPool,
                        giveaway,
                    }
                }
            ),
            tap(({ winner: { name } }) =>
                moderator.message(`@${name} has won the giveaway`)
            ),
            tap(({ winner, entryPool, giveaway }) =>
                db
                    .set(`giveaways/${entryPool}/giveaways/${giveaway}`, {
                        winners: 1,
                        slots: 1,
                    })
                    .then(() =>
                        db.set(
                            `giveaways/${entryPool}/giveaways/${giveaway}/winners/${
                                winner.id
                            }`,
                            winner
                        )
                    )
            ),
            tap(() =>
                db.set('giveaways/current', { entryPool: null, giveaway: null })
            )
        )
        .subscribe()
}
