// @rampant/moderator/index.js
import filterCommand from '@rampant/filter-command'
import makeGateOperator from '@rampant/gate-operator'
import createCommandStream from '@rampant-giveaway/commands'
import { filter, share, withLatestFrom } from 'rxjs/operators'

const cache = {}

export default ({ chat, db, streamer, prefix = '' }) => {
    const command$ = createCommandStream(chat)
    const currentCode$ = db.onDocumentChange('codes/current').pipe(share())
    const currentGiveaway$ = db
        .onDocumentChange('giveaways/current')
        .pipe(share())
    const entryChanges$ = db
        .onCollectionChange(`pools/${prefix}/entries`)
        .pipe(share())

    const intent$ = intent => {
        if (intent in cache) return cache[intent]
        const stream = command$.pipe(
            filter(({ command }) => command === intent),
            share()
        )
        cache[intent] = stream
        return stream
    }

    const gate = makeGateOperator({ chat })

    return {
        chat$: chat,
        whisper: ({ userName, message }) => chat.whisper(userName, message),
        message: chat.message,
        erase: ({ id }) => chat.erase(id),
        whisperStreamer: msg => chat.whisper(streamer, msg),
        command$: (intent, options) => {
            if (!options) return intent$(intent)
            return intent$(intent).pipe(
                gate(
                    filterCommand(options),
                    `You are not authorized to run [!${intent}]`
                )
            )
        },
        gate,
        currentCode$,
        currentGiveaway$,
        entryChanges$,
        currentCode: () => $ => $.pipe(withLatestFrom(currentCode$)),
        currentGiveaway: () => $ => $.pipe(withLatestFrom(currentGiveaway$)),
    }
}
