// @rampant-wit/plugin/index.js
import axios from 'axios'
import get from '@rampant/get'
import { mergeMap } from 'rxjs/operators'

export default ({ token } = {}) => {
    const request = ({ method, url, params, data }) =>
        axios({
            url: `https://api.wit.ai/${url}`,
            headers: { Authorization: `Bearer ${token}` },
            params: {
                v: '20170307',
                ...params,
            },
            method,
            data,
        })

    return {
        extensions: {
            nlp: ({ key, value }) =>
                mergeMap(payload => {
                    const q = get({ $: payload }, value)
                    return request({
                        method: 'get',
                        url: 'message',
                        params: { q },
                    })
                        .then(
                            ({ data }) =>
                                data.entities.intent || [{ value: 'Unknown' }]
                        )
                        .then(i => i[0].value)
                        .then(intent => ({ ...payload, [key]: intent }))
                }),
        },
    }
}
