// @rampant-twitch/rest/index.js
import twitch from 'twitch'
import admin from 'firebase-admin'

const createRestClient = (id, secret) => creds =>
    twitch.withCredentials(id, creds.access_token, creds.scope, {
        clientSecret: secret,
        expiry: creds.expires,
        onRefresh: creds.update,
        refreshToken: creds.refresh_token,
    })

export const rest = createRestClient

export default ({ clientId, clientSecret, db }) => {
    const creator = createRestClient(clientId, clientSecret)
    const cache = {}

    return userId => {
        if (userId in cache) return Promise.resolve(cache[userId])

        const update = token =>
            db.update(`users/${userId}/auth/twitch`, {
                access_token: token.accessToken,
                refresh_token: token.refreshToken,
                expires: admin.firestore.Timestamp.fromDate(token.expiryDate),
                scope: token.scope,
            })

        return db.get(`users/${userId}/auth/twitch`).then(creds => {
            const client = creator({
                ...creds,
                expires: creds.expires.toDate(),
                update,
            })
            cache[userId] = client
            return client
        })
    }
}
