// @rampant/auth/index.js
/* eslint-disable max-nested-callbacks */
import express from 'express'
import admin from 'firebase-admin'
import { decrypt } from '@dev-xp/crypto'
import functions from 'firebase-functions'
import createDb from '@rampant-firebase/db'
import { createRestClient } from '@rampant-twitch/rest'
import createTwitchAuth from '@rampant-twitch/auth'
import createTwitterAuth from '@rampant-twitter/auth'
import cookieParser from '@rampant-firebase/cookie-parser'

const twitterCreds = functions.config().twitter
const twitchCreds = functions.config().twitch
const restTwitch = createRestClient(twitchCreds)

const authTwitter = createTwitterAuth({
    ...twitterCreds,
    callback: 'https://rampant.ai/callback/twitter',
})
const authTwitch = createTwitchAuth({
    ...twitchCreds,
    callback: 'https://rampant.ai/callback/twitch',
})

const auth = express()
const authRouter = express.Router()
authRouter.use(cookieParser())

admin.initializeApp()
const firestore = admin.firestore()
const stubLogger = {
    error: () => {},
    warn: () => {},
    platform: () => {},
    plugin: () => {},
    intent: () => {},
    extend: () => {},
    detail: () => {},
    debug: () => {},
}
const logger = {
    ...stubLogger,
    child: () => stubLogger,
}
// eslint-disable-next-line no-unused-vars
const db = createDb({ db: firestore, logger })

authRouter.get('/twitch', async (req, res) => {
    // Scopes for Viewer
    // const scopes = ['chat:read', 'chat:edit', 'user_read', 'user_follows_edit'].join('+')
    // Scopes for Streamer
    const scopes = [
        'chat:read',
        'chat:edit',
        'channel:moderate',
        'user:read:email',
        'user:edit:follows',
        'channel:read:subscriptions',
        'user:edit:broadcast',
        'user:read:broadcast',
        'moderation:read',
        'clips:edit',
    ]
    res.redirect(authTwitch.getAuthUrl({ scopes, state: req.cookies }))
})

authRouter.get('/twitter', async (req, res) => {
    const authUrl = await authTwitter.getAuthUrl(req.cookies)
    res.redirect(authUrl)
})

auth.use('/auth', authRouter)
exports.auth = functions.https.onRequest(auth)

const cb = express()
const cbRouter = express.Router()

cbRouter.get('/twitter', async (req, res) => {
    const {
        state,
        oauth_token: requestToken,
        oauth_verifier: verifier,
    } = req.query
    const oldCookies = decrypt(state)
    const {
        // accessToken,
        // accessSecret,
        user: { user_id, screen_name },
    } = await authTwitter.getAccessToken({
        requestToken,
        verifier,
    })

    const cookies = {
        ...oldCookies,
        twitterId: user_id,
        twitterName: screen_name,
    }
    res.cookie('__session', JSON.stringify(cookies))
    res.redirect('/')
})

cbRouter.get('/twitch', async (req, res) =>
    authTwitch
        .getAccessToken(req.query)
        .then(({ data: credentials }) => restTwitch(credentials))
        .then(client => {
            const now = new Date()
            now.setSeconds(now.getSeconds() + client.creds.expires_in)
            return Promise.all([
                client,
                db.update('users/raybenefield', {
                    twitchId: client.user.id,
                    twitch: client.user,
                }),
                db.set('users/raybenefield/auth/twitch', {
                    ...client.creds,
                    expires: admin.firestore.Timestamp.fromDate(now),
                }),
            ])
        })
        .then(([client]) => {
            const cookies = {
                twitchId: client.user.id,
                twitchName: client.user.display_name,
            }
            res.cookie('__session', JSON.stringify(cookies))
            res.redirect('https://rampant.ai')
        })
)

cb.use('/callback', cbRouter)
exports.callback = functions.https.onRequest(cb)
