// @kikd/cli-style/index.js
import { red, green, yellow } from 'chalk'

export default {
    title: {
        display: 'block',
        color: 'bright-white',
        marginLeft: 2,
        marginBottom: 1,
    },
    check: {
        display: 'block',
        marginLeft: 4,
    },
    details: {
        color: 'grey',
        marginLeft: 4,
    },
    success: {
        display: 'block',
        color: 'grey',
        bullet: `"${green('✔')} "`,
    },
    fail: {
        display: 'block',
        color: 'red',
        bullet: `"${red('✖')} "`,
    },
    warning: {
        display: 'block',
        color: 'bright-yellow',
        bullet: `"${yellow('⚠')} "`,
    },
    list: {
        display: 'block',
        marginTop: 1,
    },
    item: {
        display: 'block',
        marginBottom: 1,
    },
    bullet: {
        display: 'block',
        marginLeft: 2,
    },
    change: { display: 'block' },
    type: { color: 'white' },
    scope: { color: 'bright-white' },
    subject: { color: 'grey' },
    attr: {
        display: 'block',
        marginLeft: 2,
    },
    key: {
        color: 'grey',
        marginRight: 1,
    },
    value: { color: 'white' },
    bar: {
        display: 'block',
        color: 'green',
        bullet: `"${green(
            '────────────────────────────────────────────────────────────────────────────────'
        )}"`,
    },
    h1: {
        display: 'block',
        color: 'bright-white',
        marginLeft: 2,
    },
}
