// @kikd/command-build/index.js
import { isolate } from 'transmutation'
import template from '@kikd/template-building'
import getPackages from '@kikd/command-list'
import buildPackages from '@kikd/build-packages'
import getDependencies from '@kikd/command-deps'
import buildRollupConfigs from '@kikd/build-rollup-configs'

export default function build(config) {
    const {
        allPackages = false,
        changedAndPublishable = false,
        needsPublishing = false,
    } = config
    const newConfig = {
        ...config,
        ...(!allPackages && !changedAndPublishable && !needsPublishing
            ? { onlyPublishable: true }
            : {}),
    }

    return getPackages(newConfig)
        .extend(
            'dependencies',
            getDependencies(newConfig).then(isolate('dependencies'))
        )
        .extend('configs', buildRollupConfigs(config))
}

export const command = () => ({
    name: 'build',
    run: build,
    template,
    effect: buildPackages,
    args: {
        _: 'packageNames',
        changed: 'changedAndPublishable',
        publishing: 'needsPublishing',
    },
})
