// @kikd/command-build/index.js
import { isolate } from 'transmutation'

import getPackages from '@kikd/command-list'
import npmCli from '@kikd/plugin-build-npm-cli'
import entyrePackage from '@entyre/kikd-plugin'
import buildPackages from '@kikd/build-packages'
import getDependencies from '@kikd/command-deps'
import vueSite from '@kikd/plugin-build-vue-site'
import npmPackage from '@kikd/plugin-build-npm-package'
import buildRollupConfigs from '@kikd/build-rollup-configs'

const plugins = [npmPackage, npmCli, vueSite, entyrePackage].reduce(
    (all, plugin) => ({ ...all, [plugin.type]: plugin }),
    {}
)

export default function build(config) {
    return getPackages({
        ...config,
        ...(!config.allPackages && !config.changedAndPublishable
            ? { onlyPublishable: true }
            : {}),
    })
        .extend(
            'dependencies',
            getDependencies(config).then(isolate('dependencies'))
        )
        .extend('configs', buildRollupConfigs(plugins))
}

export const command = () => ({
    name: 'build',
    renderer: () => {},
    run: build,
    effect: buildPackages(plugins),
    args: {
        _: 'packageNames',
        all: 'allPackages',
        changed: 'changedAndPublishable',
    },
})
