// @semcom/relevant-diffs/snapshot-tests.js
import describe from 'kape'
import relevantDiffs from '@semcom/relevant-diffs'

const diffs = [
    {
        hunks: [
            {
                content: '@@ -32,6 +32,14 @@ define-command sort %{',
                oldStart: 32,
                newStart: 32,
                oldLines: 6,
                newLines: 14,
                changes: [
                    {
                        content: `    execute-keys "|awk '{ print length, $0 }' | sort -n | cut -d' ' -f2-<ret>"`,
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 32,
                        newLineNumber: 32,
                    },
                    {
                        content: '}',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 33,
                        newLineNumber: 33,
                    },
                    {
                        content: '',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 34,
                        newLineNumber: 34,
                    },
                    {
                        content: 'define-command test %{',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 35,
                    },
                    {
                        content: '    evaluate-commands %{',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 36,
                    },
                    {
                        content: '        edit %sh{',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 37,
                    },
                    {
                        content: `            echo $kak_buffile | xargs dirname | awk '{print $1"/snapshot-tests.js"}'`,
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 38,
                    },
                    {
                        content: '        }',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 39,
                    },
                    {
                        content: '    }',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 40,
                    },
                    {
                        content: '}',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 41,
                    },
                    {
                        content: '',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 42,
                    },
                    {
                        content:
                            'define-command -docstring "map-sequence <scope> <sequence> <command>: map <sequence> of keys to <command> in insert mode." \\',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 35,
                        newLineNumber: 43,
                    },
                    {
                        content:
                            'map-sequence -params 3 %{ evaluate-commands %sh{',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 36,
                        newLineNumber: 44,
                    },
                    {
                        content:
                            '    keys=$(printf "%s" "$2" | sed "s/\\([&|]\\)/\\1\\1/g")',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 37,
                        newLineNumber: 45,
                    },
                ],
            },
        ],
        oldEndingNewLine: true,
        newEndingNewLine: true,
        oldRevision: 'fbcb3585',
        newRevision: '35233601',
        newMode: '100644',
        oldMode: '100644',
        oldPath: '.files/kak/kakrc',
        newPath: '.files/kak/kakrc',
        type: 'modify',
    },
    {
        hunks: [
            {
                content: '@@ -122,6 +122,7 @@',
                oldStart: 122,
                newStart: 122,
                oldLines: 6,
                newLines: 7,
                changes: [
                    {
                        content: '        "firebase-admin": "^8.2.0",',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 122,
                        newLineNumber: 122,
                    },
                    {
                        content: '        "front-matter": "^2.3.0",',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 123,
                        newLineNumber: 123,
                    },
                    {
                        content: '        "fuzzy": "^0.1.3",',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 124,
                        newLineNumber: 124,
                    },
                    {
                        content: '        "gitdiff-parser": "^0.2.2",',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 125,
                    },
                    {
                        content: '        "glob": "^7.1.2",',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 125,
                        newLineNumber: 126,
                    },
                    {
                        content: '        "inquirer": "^7.0.0",',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 126,
                        newLineNumber: 127,
                    },
                    {
                        content:
                            '        "inquirer-autocomplete-prompt": "^0.12.1",',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 127,
                        newLineNumber: 128,
                    },
                ],
            },
        ],
        oldEndingNewLine: true,
        newEndingNewLine: true,
        oldRevision: '762de6cc',
        newRevision: 'b427a4be',
        newMode: '100644',
        oldMode: '100644',
        oldPath: 'package.json',
        newPath: 'package.json',
        type: 'modify',
    },
    {
        hunks: [
            {
                content: '@@ -1,10 +1,14 @@',
                oldStart: 1,
                newStart: 1,
                oldLines: 10,
                newLines: 14,
                changes: [
                    {
                        content: "import log from '@dev-xp/log'",
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 1,
                        newLineNumber: 1,
                    },
                    {
                        content: "import diffParser from 'gitdiff-parser'",
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 2,
                    },
                    {
                        content: "import { sync as root } from '@dev-xp/root'",
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 2,
                        newLineNumber: 3,
                    },
                    {
                        content:
                            "import configureGit from 'simple-git/promise'",
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 3,
                        newLineNumber: 4,
                    },
                    {
                        content:
                            "import patchObject from '@dev-xp/patch-object'",
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 4,
                        newLineNumber: 5,
                    },
                    {
                        content:
                            "import camelCaseKeys from '@dev-xp/camel-case-keys'",
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 5,
                        newLineNumber: 6,
                    },
                    {
                        content: '',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 6,
                        newLineNumber: 7,
                    },
                    {
                        content:
                            'const patch = patchObject({ status: camelCaseKeys })',
                        type: 'delete',
                        isDelete: true,
                        lineNumber: 7,
                    },
                    {
                        content: 'const patch = patchObject({',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 8,
                    },
                    {
                        content: '    status: camelCaseKeys,',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 9,
                    },
                    {
                        content:
                            '    diff: result => log.deep(diffParser.parse(result)),',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 10,
                    },
                    {
                        content: '})',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 11,
                    },
                    {
                        content:
                            'const makeGit = () => patch(configureGit(root()))',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 8,
                        newLineNumber: 12,
                    },
                    {
                        content: '',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 9,
                        newLineNumber: 13,
                    },
                    {
                        content: 'export const silent = makeGit()',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 10,
                        newLineNumber: 14,
                    },
                ],
            },
        ],
        oldEndingNewLine: true,
        newEndingNewLine: true,
        oldRevision: '5e7d2afa',
        newRevision: '576d7eaa',
        newMode: '100644',
        oldMode: '100644',
        oldPath: 'src/node_modules/@dev-xp/git/index.js',
        newPath: 'src/node_modules/@dev-xp/git/index.js',
        type: 'modify',
    },
    {
        hunks: [
            {
                content: '@@ -1,8 +1,11 @@',
                oldStart: 1,
                newStart: 1,
                oldLines: 8,
                newLines: 11,
                changes: [
                    {
                        content: "import log from '@dev-xp/log'",
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 1,
                    },
                    {
                        content: "import gitStatus from '@semcom/git-status'",
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 1,
                        newLineNumber: 2,
                    },
                    {
                        content: "import { silent as git } from '@dev-xp/git'",
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 2,
                        newLineNumber: 3,
                    },
                    {
                        content:
                            "import resetVsAdd from '@semcom/reset-vs-add'",
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 3,
                        newLineNumber: 4,
                    },
                    {
                        content:
                            "import relevantDiffs from '@semcom/relevant-diffs'",
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 5,
                    },
                    {
                        content:
                            "import requestCommit from '@semcom/request-commit'",
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 4,
                        newLineNumber: 6,
                    },
                    {
                        content:
                            "import transmute, { isolate } from 'transmutation'",
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 5,
                        newLineNumber: 7,
                    },
                    {
                        content:
                            "import filesToScopes from '@semcom/files-to-scopes'",
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 8,
                    },
                    {
                        content: '',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 6,
                        newLineNumber: 9,
                    },
                    {
                        content: 'export default function status() {',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 7,
                        newLineNumber: 10,
                    },
                    {
                        content: '    return transmute({})',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 8,
                        newLineNumber: 11,
                    },
                ],
            },
            {
                content:
                    '@@ -14,6 +17,12 @@ export default function status() {',
                oldStart: 14,
                newStart: 17,
                oldLines: 6,
                newLines: 12,
                changes: [
                    {
                        content:
                            "        .extend('resetVsAdd', ({ statusFiles, filesToCommit }) =>",
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 14,
                        newLineNumber: 17,
                    },
                    {
                        content:
                            '            resetVsAdd({ statusFiles, filesToCommit })',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 15,
                        newLineNumber: 18,
                    },
                    {
                        content: '        )',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 16,
                        newLineNumber: 19,
                    },
                    {
                        content: "        .extend('diffs', () => git.diff())",
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 20,
                    },
                    {
                        content:
                            "        .extend('diffsForCommit', relevantDiffs())",
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 21,
                    },
                    {
                        content:
                            "        .log('diffsForCommit', 'diffsForCommit', log.deep)",
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 22,
                    },
                    {
                        content:
                            "        .extend('scopedChanges', ({ filesToCommit }) =>",
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 23,
                    },
                    {
                        content: '            filesToScopes(filesToCommit)',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 24,
                    },
                    {
                        content: '        )',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 25,
                    },
                    {
                        content: '}',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 17,
                        newLineNumber: 26,
                    },
                    {
                        content: '',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 18,
                        newLineNumber: 27,
                    },
                    {
                        content: 'export const command = () => ({',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 19,
                        newLineNumber: 28,
                    },
                ],
            },
        ],
        oldEndingNewLine: true,
        newEndingNewLine: true,
        oldRevision: '114a3b9b',
        newRevision: '26a2c16d',
        newMode: '100644',
        oldMode: '100644',
        oldPath: 'src/node_modules/@semcom/command-commit/index.js',
        newPath: 'src/node_modules/@semcom/command-commit/index.js',
        type: 'modify',
    },
    {
        hunks: [
            {
                content: '@@ -3410,6 +3410,11 @@ git-raw-commits@^1.3.0:',
                oldStart: 3410,
                newStart: 3410,
                oldLines: 6,
                newLines: 11,
                changes: [
                    {
                        content: '    split2 "^2.0.0"',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 3410,
                        newLineNumber: 3410,
                    },
                    {
                        content: '    through2 "^2.0.0"',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 3411,
                        newLineNumber: 3411,
                    },
                    {
                        content: '',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 3412,
                        newLineNumber: 3412,
                    },
                    {
                        content: 'gitdiff-parser@^0.2.2:',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 3413,
                    },
                    {
                        content: '  version "0.2.2"',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 3414,
                    },
                    {
                        content:
                            '  resolved "https://registry.yarnpkg.com/gitdiff-parser/-/gitdiff-parser-0.2.2.tgz#c393ce9caf385860d005ce758a5e03221a7fc6be"',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 3415,
                    },
                    {
                        content:
                            '  integrity sha512-iUE5gWohbUm+hogg9c1Gg/pDoDyGlC15hTdsnW7AqZw9+F9mYAxfnMNDstFxuF+qjiZot2s+3L+K9CLdw1ttvA==',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 3416,
                    },
                    {
                        content: '',
                        type: 'insert',
                        isInsert: true,
                        lineNumber: 3417,
                    },
                    {
                        content: 'glob-base@^0.3.0:',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 3413,
                        newLineNumber: 3418,
                    },
                    {
                        content: '  version "0.3.0"',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 3414,
                        newLineNumber: 3419,
                    },
                    {
                        content:
                            '  resolved "https://registry.npmjs.org/glob-base/-/glob-base-0.3.0.tgz#dbb164f6221b1c0b1ccf82aea328b497df0ea3c4"',
                        type: 'normal',
                        isNormal: true,
                        oldLineNumber: 3415,
                        newLineNumber: 3420,
                    },
                ],
            },
        ],
        oldEndingNewLine: true,
        newEndingNewLine: true,
        oldRevision: '1e4b46a3',
        newRevision: '7ec88154',
        newMode: '100644',
        oldMode: '100644',
        oldPath: 'yarn.lock',
        newPath: 'yarn.lock',
        type: 'modify',
    },
]

describe('@semcom/relevant-diffs', relevantDiffs(), snapshot => {
    snapshot('Relevant Diffs', [
        [],
        [{ filesToCommit: ['package.json'], diffs }],
    ])
})
