// @dev-xp/fs/index.js
import {
    rmSync,
    readFile,
    mkdirSync,
    writeFile,
    existsSync,
    readdirSync,
    readFileSync,
    writeFileSync,
    createReadStream,
    lstat as lstatOg,
} from 'fs'

import globOg from 'glob'
import promisify from 'es6-promisify'
import jsonFunctions from '@dev-xp/json'
import { copy as xCopy } from 'fs-extra'

export const readStream = createReadStream
export const mkdir = path => mkdirSync(path, { recursive: true })
export const rmdir = dir => rmSync(dir, { recursive: true, force: true })
export const readDir = readdirSync
export const exists = existsSync
export const readSync = readFileSync
export const writeSync = writeFileSync

export const lstat = promisify(lstatOg)
export const copy = promisify(xCopy)
export const read = promisify(readFile)
export const write = promisify(writeFile)

export const glob = promisify(globOg)

export const json = jsonFunctions

const some = { readStream, mkdir, readDir, exists, readSync, writeSync }
const fs = { ...some, lstat, copy, read, write, glob, json }

export default fs
