// @kikd/build-rollup-configs/index.js
import { minify } from 'uglify-es';

import builtins from 'builtin-modules';
import babel from 'rollup-plugin-babel';
import uglify from 'rollup-plugin-uglify';
import filesize from 'rollup-plugin-filesize';
import executable from 'rollup-plugin-executable';
import resolve from 'rollup-plugin-node-resolve';

const babelConfig = {
    babelrc: false,
    presets: [['env', { modules: false, targets: { node: '6' } }], 'stage-0'],
    plugins: ['external-helpers'],
    exclude: ['node_modules/**'],
};

const basePackageInfo = pkg => ({
    name: pkg.name,
    srcDir: `./src/node_modules/${pkg.name}`,
    destDir: `./dist/${pkg.name}`,
    packageConfig: pkg.config,
    input: `./src/node_modules/${pkg.name}/${pkg.file}`,
});

const plugins = [
    resolve({
        customResolveOptions: {
            moduleDirectory: 'src/node_modules',
        },
    }),
    babel(babelConfig),
    uglify({}, minify),
    filesize(),
];

const binConfigs = ({ packages, baseConfig }) =>
    packages
        .filter(p => p.bin)
        .filter(p => p.bin.constructor === Object)
        .map(pkg =>
            Object.values(pkg.bin).map(file => ({
                ...basePackageInfo({ ...pkg, file }),
                external: [
                    ...builtins,
                    ...Object.keys(baseConfig.dependencies),
                ],
                output: [
                    {
                        file: `./dist/${pkg.name}/${file}`,
                        format: 'cjs',
                    },
                ],
                banner: '#!/usr/bin/env node',
                plugins: [...plugins, executable()],
            })),
        );

const mainConfigs = ({ packages, baseConfig }) =>
    packages
        .filter(p => !p.bin)
        .map(p => ({
            ...p,
            file: p.main || 'index.js',
        }))
        .map(pkg => ({
            ...basePackageInfo(pkg),
            external: [...builtins, ...Object.keys(baseConfig.dependencies)],
            output: [
                {
                    file: `./dist/${pkg.name}/${pkg.file}`,
                    format: 'cjs',
                },
            ],
            plugins,
        }));

export default ({ packages = [], baseConfig = { dependencies: {} } } = {}) =>
    new Promise(res =>
        res(
            [].concat(
                ...binConfigs({ packages, baseConfig }),
                ...mainConfigs({ packages, baseConfig }),
            ),
        ),
    );
