// @rampant-wit/api/index.js
/* eslint-disable no-console */
import axios from 'axios'

export default token => {
    const request = ({ method, url, params, data }) =>
        axios({
            url: `https://api.wit.ai/${url}`,
            headers: { Authorization: `Bearer ${token}` },
            params: {
                v: '20170307',
                ...params,
            },
            method,
            data,
        })
            .then(({ data: responseData }) => responseData)
            .catch(res => {
                console.log(`method: ${method}`)
                console.log(`url: ${url}`)
                console.log(`params: ${JSON.stringify(params)}`)
                console.log(`data: ${JSON.stringify(data)}`)
                console.log(`response data: ${res.data}`)
            })

    return {
        analyze: q => request({ method: 'get', url: 'message', params: { q } }),
        trainSamples: (samples, intent) =>
            request({
                method: 'post',
                url: `samples`,
                data: samples.map(text => ({
                    text,
                    entities: intent
                        ? [{ entity: 'intent', value: intent }]
                        : [],
                })),
            }),
    }
}
