// @rampant-wit/api/index.js
import axios from 'axios'
import { prettify } from '@rampant/logger'

export default ({ token, logger }) => {
    const request = ({ method, url, params, data }) =>
        axios({
            url: `https://api.wit.ai/${url}`,
            headers: { Authorization: `Bearer ${token}` },
            params: {
                v: '20170307',
                ...params,
            },
            method,
            data,
        })
            .then(({ data: responseData }) => responseData)
            .then(responseData => {
                logger.debug(`Got results from [${method} api.wit.ai/${url}]`)
                logger.debug(prettify({ params, data }))
                logger.debug('Response:')
                logger.debug(prettify(responseData))
                return responseData
            })
            .catch(res => {
                logger.error('Wit API call failed')
                logger.error(
                    prettify({
                        method,
                        url,
                        params,
                        data,
                        responseData: res.data,
                        responseStatus: res.status,
                    })
                )
            })

    return {
        analyze: q => request({ method: 'get', url: 'message', params: { q } }),
        getSamples: intent =>
            request({
                method: 'get',
                url: `samples`,
                params: {
                    entity_ids: ['intent'],
                    entity_values: [intent],
                    limit: 10000,
                },
            }),
        trainSamples: (samples, intent) =>
            request({
                method: 'post',
                url: `samples`,
                data: samples.map(text => ({
                    text,
                    entities: intent
                        ? [{ entity: 'intent', value: intent }]
                        : [],
                })),
            }),
    }
}
