// @fever/moderator/index.js
import filterCommand from '@fever/filter-command'
import { tap, filter, share, withLatestFrom } from 'rxjs/operators'
import createCommandStream from '@rampant-giveaway/commands'

const cache = {}

export default ({ chat, db, streamer }) => {
    const command$ = createCommandStream(chat)
    const currentCode$ = db.onDocumentChange('codes/current')
    const currentGiveaway$ = db.onDocumentChange('giveaways/current')

    const intent$ = intent => {
        if (intent in cache) return cache[intent]
        const stream = command$.pipe(
            filter(({ command }) => command === intent),
            share()
        )
        cache[intent] = stream
        return stream
    }

    const gate = (gateFilter, message) => stream$ =>
        stream$.pipe(
            tap(args => {
                const user = Array.isArray(args)
                    ? args[0].userName
                    : args.userName
                const msg =
                    typeof message === 'function' ? message(args) : message

                if (!gateFilter(args)) chat.whisper(user, msg)
            }),
            filter(gateFilter)
        )

    return {
        whisper: ({ userName, message }) => chat.whisper(userName, message),
        message: chat.message,
        whisperStreamer: msg => chat.whisper(streamer, msg),
        command$: (intent, options) => {
            if (!options) return intent$(intent)
            return intent$(intent).pipe(
                gate(
                    filterCommand(options),
                    `You are not authorized to run [!${intent}]`
                )
            )
        },
        gate,
        currentCode$,
        currentGiveaway$,
        currentCode: $ => $.pipe(withLatestFrom(currentCode$)),
        currentGiveaway: $ => $.pipe(withLatestFrom(currentGiveaway$)),
    }
}
