// @runic-unreal/blueprint/index.js
import log from '@dev-xp/log'
import { runes } from 'runic'
import * as _ from '@dev-xp/utils'
import tile from '@runic-unreal/tile'
import AreaPlugin from '@sigils/area-plugin'
import VueRenderPlugin from 'rete-vue-render-plugin'
import ConnectionPlugin from 'rete-connection-plugin'
import ContextMenuPlugin from 'rete-context-menu-plugin'

export { default as table } from '@runic-unreal/table'

export default ({ background } = {}) => ({
    tile,
    runes: {
        number: runes.number(),
    },
    alterations: {
        ConnectionPlugin,
        VueRenderPlugin,
        AreaPlugin: [AreaPlugin, { background }],
        ContextMenuPlugin: [
            ContextMenuPlugin,
            {
                searchBar: false,
                delay: 1000,
                allocate(component) {
                    return [_.cond([[_.T, _.always('Nodes')]])(component)]
                },
                rename(component) {
                    return component.name
                },
                items: {
                    // Example: () => console.log('Works!'),
                },
            },
        ],
    },
})
