// @runic-unreal/blueprint/index.js
import log from '@dev-xp/log'
import { runes } from 'runic'
import * as _ from '@dev-xp/utils'
import tile from '@runic-unreal/tile'
import { capitalCase } from 'change-case'
import AreaPlugin from '@runic/area-plugin'
import VueRenderPlugin from 'rete-vue-render-plugin'
import ConnectionPlugin from 'rete-connection-plugin'
import ContextMenuPlugin from 'rete-context-menu-plugin'

export { default as table } from '@runic-unreal/table'

export default ({ background } = {}) => ({
    tile,
    runes,
    alterations: {
        ConnectionPlugin,
        VueRenderPlugin,
        AreaPlugin: [AreaPlugin, { background }],
        ContextMenuPlugin: [
            ContextMenuPlugin,
            {
                searchBar: true,
                delay: 100,
                allocate(component) {
                    const fullPath = _.pipe(
                        _.prop('path'),
                        _.split('.'),
                        _.init,
                        _.map(capitalCase),
                        _.prepend('Nodes')
                    )

                    return _.cond([[_.T, fullPath]])(component)
                },
                rename(component) {
                    return component.name
                },
                items: {
                    // Example: () => console.log('Works!'),
                },
            },
        ],
    },
})
