// @kikd/command-list/index.js
import transmute from 'transmutation'
import findRepoRoot from '@dev-xp/root'
import getTypes from '@kikd/package-types'
import extraDetails from '@kikd/extra-details'
import getRootManifest from '@kikd/root-manifest'
import filterPackages from '@kikd/filter-packages'
import template from '@kikd/template-package-list'
import getRootStructure from '@kikd/root-structure'
import getStructures from '@kikd/package-structure'
import getPackageDetails from '@kikd/package-details'

export default function list(config) {
    return transmute({})
        .extend('root', findRepoRoot)
        .extend('rootStructure', getRootStructure)
        .extend('rootManifest', getRootManifest)
        .extend('exclude', ['**/fixtures/**'])
        .extend('packageNames', filterPackages(config))
        .extend('structures', getStructures)
        .extend('types', getTypes(config))
        .extend('manifests', getPackageDetails(config))
        .extendEach('details', 'packageNames', extraDetails(config))
}

export const packageFilters = {
    _: 'packageNames',
    all: 'allPackages',
    publishable: 'onlyPublishable',
    changed: 'onlyChanged',
    releasing: 'changedAndPublishable',
    publishing: 'needsPublishing',
}

export const command = () => ({
    name: 'list',
    aliases: ['ls'],
    run: list,
    template,
    args: packageFilters,
})
