// @kikd/plugin-build-vue-site/index.js
import fs from 'fs'
import { resolve } from 'path'

import promisify from 'es6-promisify'
import build from '@kikd/build-rollup'
import postBuild from '@kikd/post-build'
import { input, output, postInfo, commonPlugins } from '@kikd/rollup-utils'
import commonjs from 'rollup-plugin-commonjs'
import globals from 'rollup-plugin-node-globals'
import vuePlugin from 'rollup-plugin-vue'

const copy = promisify(fs.copyFile)
const type = 'vue-site'
const plugins = [
    vuePlugin({ css: true }),
    commonjs(),
    globals(),
    ...commonPlugins,
]

export default {
    type,
    filter: pkg =>
        pkg.files.includes('index.html') && pkg.files.includes('app.js'),
    config: ({ baseConfig, dependencies }) => p => {
        const pkg = { ...p, file: 'app.js' }
        return {
            type,
            in: input({ pkg, dependencies, plugins }),
            out: [output({ format: 'iife', pkg })],
            post: postInfo({ pkg, baseConfig, dependencies }),
        }
    },
    build,
    postBuild: ({ root, baseConfig, config }) =>
        postBuild({ root, baseConfig, config }).then(() =>
            copy(
                resolve(root, config.srcDir, 'index.html'),
                resolve(root, config.destDir, 'index.html')
            )
        ),
}
