// @kikd/plugin-build-vue-site/index.js
import fs from 'fs';
import { resolve } from 'path';
import promisify from 'es6-promisify';
import build from '@kikd/build-rollup';
import postBuild from '@kikd/post-build';
import { input, output, postInfo, plugins } from '@kikd/rollup-utils';

const copy = promisify(fs.copyFile);
const type = 'vue-site';
export default {
    type,
    filter: pkg =>
        pkg.files.includes('index.html') && pkg.files.includes('app.js'),
    config: ({ baseConfig, dependencies }) => p => {
        const pkg = { ...p, file: 'app.js' };
        return {
            type,
            in: input({ pkg, dependencies, plugins: plugins({ vue: true }) }),
            out: [output({ format: 'iife', pkg })],
            post: postInfo({ pkg, baseConfig, dependencies }),
        };
    },
    build,
    postBuild: ({ root, baseConfig, config }) =>
        postBuild({ root, baseConfig, config }).then(() =>
            copy(
                resolve(root, config.srcDir, 'index.html'),
                resolve(root, config.destDir, 'index.html'),
            ),
        ),
};
