// @entyre/kikd-plugin/index.js
import { extname } from 'path'
import build from '@kikd/build-rollup'
import entyre from '@entyre/rollup-plugin'
import { input, output, postInfo, commonPlugins } from '@kikd/rollup-utils'

const type = 'entyre-package'

export default {
    type,
    filter: pkg =>
        (pkg.main && extname(pkg.main) === '.md') ||
        pkg.files.includes('index.md'),
    config: ({ baseConfig, dependencies }) => p => {
        const pkg = { ...p, file: p.main || 'index.md' }
        return {
            type,
            in: input({
                pkg,
                dependencies,
                plugins: [
                    entyre({
                        readmePath: `./dist/${pkg.name}/readme.md`,
                        packageJsonPath: `./dist/${pkg.name}/package.json`,
                    }),
                    ...commonPlugins,
                ],
            }),
            out: [
                output({
                    pkg: { ...pkg, file: 'index.js' },
                    outFile: 'index.js',
                    format: 'cjs',
                }),
            ],
            post: postInfo({ pkg, baseConfig, dependencies }),
        }
    },
    build,
    postBuild: () => {},
}
