// @rampant/ai/test-intents.js
import { green, grey } from 'chalk'

const success = msg => `    ${green('✔')} ${grey(msg)}`

export const intents = [
    {
        name: 'logTest',
        sources: [{}],
        defaults: { message: success('Log effects work') },
        effects: [{ log: '{params.message}' }],
    },
    {
        name: 'vote',
        sources: [{}],
        defaults: {
            doc: 'votes',
            data: { dusty: 1 },
            message: success('Database effects work'),
        },
        effects: [
            { db: { doc: '{params.doc}', data: '{params.data}' } },
            { log: '{params.message}' },
        ],
    },
    {
        name: 'voteResults',
        sources: [{ db: 'votes' }],
        defaults: {
            doc: 'vote-results',
            data: { dusty: 1 },
            message: success('Database sources work'),
        },
        effects: [
            { db: { doc: '{params.doc}', data: '{params.data}' } },
            { log: '{params.message}' },
        ],
    },
]

export const cleanupDocs = ['votes', 'vote-results']
