// @semcom/command-prerelease/index.js
import effect from '@semcom/prerelease-package'
import template from '@semcom/template-prerelease'
import pickPackage from '@kikd/prompt-pick-package'
import getCheckedPackages from '@kikd/command-check'
import getStructures from '@kikd/package-structure'
import fixPackages from '@semcom/fix-packages'

export default function prerelease(config) {
    return getCheckedPackages(config)
        .extend('unpublishable', ({ packageNames, checks }) =>
            packageNames.filter(name => !checks.packages[name].publishable)
        )
        .extend('name', ({ unpublishable }) =>
            pickPackage({
                message: 'Which package should we prerelease?',
                packageNames: unpublishable,
            })
        )
        .extend(
            'structure',
            ({ name }) => getStructures({ packageNames: [name] })[name]
        )
        .extend(fixPackages)
}

export const command = () => ({
    name: 'prerelease',
    run: prerelease,
    template,
    effect,
})
