// @kikd/build-packages/index.js
import rimraf from 'rimraf';
import cpx from 'cpx';
import { resolve } from 'path';
import { rollup } from 'rollup';
import { promisify } from 'util';

const removeDir = promisify(rimraf);
const copy = promisify(cpx.copy);

export default ({ root, configs = [] } = {}) =>
    removeDir(resolve(root, 'dist')).then(() =>
        Promise.all(
            configs.map(config =>
                rollup({
                    input: config.input,
                    external: config.external,
                    plugins: config.plugins,
                })
                    .then(bundle =>
                        Promise.all(
                            config.output.map(out =>
                                bundle.write({
                                    file: out.file,
                                    format: out.format,
                                    banner: config.banner,
                                }),
                            ),
                        ),
                    )
                    .then(() =>
                        Promise.all([
                            copy(
                                resolve(root, 'LICENSE'),
                                resolve(config.destDir),
                                { dereference: true },
                            ),
                            copy(
                                resolve(config.srcDir, 'readme.md'),
                                resolve(config.destDir),
                                { dereference: true },
                            ),
                        ]),
                    ),
            ),
        ),
    );
