// @kikd/build-packages/index.js
import rimraf from 'rimraf';
import { resolve } from 'path';
import { rollup } from 'rollup';
import promisify from 'es6-promisify';
import postBuild from '@kikd/post-build';

const removeDir = promisify(rimraf);

export default ({ root, baseConfig, configs = [] } = {}) =>
    removeDir(resolve(root, 'dist')).then(() =>
        Promise.all(
            configs.map(config =>
                rollup({
                    input: resolve(root, config.input),
                    external: config.external,
                    plugins: config.plugins,
                })
                    .then(bundle =>
                        Promise.all(
                            config.output.map(out =>
                                bundle.write({
                                    file: resolve(root, out.file),
                                    format: out.format,
                                    banner: config.banner,
                                    exports:
                                        bundle.exports.length > 1 // eslint-disable-line no-nested-ternary
                                            ? 'named'
                                            : bundle.exports.length === 0 // eslint-disable-line no-nested-ternary
                                              ? 'none'
                                              : bundle.exports[0] === 'default'
                                                ? 'default'
                                                : 'named',
                                }),
                            ),
                        ),
                    )
                    .then(postBuild({ root, baseConfig, config })),
            ),
        ),
    );
