// @coregames/parse-pbt/index.js
/* eslint-disable no-unused-vars */
import p from 'parsimmon'
import int from '@dev-xp/parse-int'
import isObject from 'lodash.isobject'

const getNum = index => arr => arr[index]

const language = p.createLanguage({
    lbrace: r => p.string('{'),
    rbrace: r => p.string('}'),
    colon: r => p.string(':'),
    space: r => p.string(' '),
    LF: r => p.string('\n'),
    CR: r => p.string('\r\n'),
    CRLF: r => p.seq(r.CR, r.LF),
    EOF: r => p.alt(r.LF, r.CRLF),
    key: r =>
        p.seq(r.whitespace, r.identifier, r.colon, r.space).map(getNum(1)),
    whitespace: r => p.regexp(/\s*/i),
    identifier: r => p.regexp(/[a-z_][a-z0-9_]*/i),
    chars: r => p.regexp(/[a-z0-9_:-\s\\']*/i),
    quote: r => p.string('"'),
    string: r => r.chars.wrap(r.quote, r.quote),
    number: r =>
        p.regexp(/[0-9.\-e]*/i).map(number => Math.round(number * 100) / 100),
    true: r => p.string('true'),
    false: r => p.string('false'),
    boolean: r => p.alt(r.true, r.false).map(x => x === 'true'),
    server: r => p.string('Server'),
    static: r => p.string('RuntimeStatic'),
    enums: r => p.alt(r.server, r.static),
    values: r => p.alt(r.boolean, r.enums, r.string, r.number),
    property: r => p.seqObj(['key', r.key], ['value', r.values]),
    objStart: r => p.seq(r.whitespace, r.identifier, r.space).map(getNum(1)),
    objValues: r => p.alt(r.obj, r.property).many(),
    objEnd: r => p.seq(r.whitespace, r.rbrace),
    obj: r =>
        p.seqObj(
            ['key', r.objStart],
            ['value', r.objValues.trim(r.whitespace).wrap(r.lbrace, r.rbrace)]
        ),
    root: r => r.objValues.skip(r.EOF),
})

export default unparsed => (!unparsed ? [] : language.root.tryParse(unparsed))
