// @rampant/pipeline-creator/index.js
import { map, tap } from 'rxjs/operators'
import templating from '@rampant/templating'

export default () => pipe => {
    const parseTypes = {
        number: val => parseInt(val, 10),
        object: val => JSON.parse(val),
        string: val => val,
    }
    const extensionTypes = {
        static: ({ key, value, parse = 'string' }) => {
            const render = templating.template(value)

            return map(payload => {
                const preprocessed = render(payload.context)
                const processor = parseTypes[parse]
                const processed = processor(preprocessed)

                return {
                    ...payload,
                    context: {
                        ...payload.context,
                        [key]: processed,
                    },
                }
            })
        },
        blank: () => tap(() => {}),
    }

    const { type } = pipe
    const extensionCreator = extensionTypes[type]
    return extensionCreator(pipe)
}
