// @kikd/create-cli/index.js
/* eslint max-lines: ['warn', { max: 52, skipBlankLines: true }] */
/* eslint-disable no-console */
import { join } from 'path'

import dedent from 'dedent'
import { prompt } from 'inquirer'
import { write, mkdir } from '@dev-xp/fs'
import { edit, writeSync as writeJson } from '@dev-xp/json'

export const preCreate = ({ packagePath, name }) =>
    prompt([
        {
            type: 'input',
            name: 'executableName',
            message: 'What should the executable name be?',
            default: name,
        },
        {
            type: 'editor',
            name: 'executableContents',
            message: 'What should the initial contents be for the executable?',
            default: dedent(`
                import log from '@dev-xp/log'

                log('${name}')
            `),
        },
    ]).then(({ executableName, executableContents }) => ({
        cliPath: join(packagePath, 'cli.js'),
        manifestPath: join(packagePath, 'package.json'),
        manifestContents: { bin: { [executableName]: 'cli.js' } },
        executableName,
        executableContents,
        scriptPath: `scripts.${executableName}`,
        script: `cd src/node_modules && babel-node ${name}/cli.js`,
    }))

export const createTemplate = ({ cliPath, executableName, script }) => `
    <attr><key>CLI Path:</key><value>${cliPath}</value></attr>
    <attr><key>Executable Name:</key><value>${executableName}</value></attr>
    <attr><key>Run Script:</key><value>${script}</value></attr>
`

export default ({
    rootManifestPath,
    packagePath,
    extra: {
        cliPath,
        manifestPath,
        manifestContents,
        executableContents,
        scriptPath,
        script,
    } = {},
} = {}) => {
    mkdir(packagePath)
    writeJson(manifestPath, manifestContents)
    write(cliPath, executableContents)
    edit(rootManifestPath, scriptPath, script)
}
