// @portfolio/app/portfolio.js
/* eslint-disable max-lines */
export default {
    id: 'portfolio@0.0.0',
    nodes: {
        1: {
            id: 1,
            data: { num: 1 },
            inputs: {},
            outputs: {
                num: {
                    connections: [{ node: 4, input: 'a', data: {} }],
                },
            },
            position: [0, -122],
            name: 'Number',
        },
        2: {
            id: 2,
            data: { num: 2 },
            inputs: {},
            outputs: { num: { connections: [] } },
            position: [-1.4878699295430766, 119.99996948242188],
            name: 'Number',
        },
        3: {
            id: 3,
            data: { num: 3 },
            inputs: {},
            outputs: {
                num: {
                    connections: [{ node: 4, input: 'b', data: {} }],
                },
            },
            position: [0, 0],
            name: 'Number',
        },
        4: {
            id: 4,
            data: { a: 5, b: 7 },
            inputs: {
                a: {
                    connections: [{ node: 1, output: 'num', data: {} }],
                },
                b: {
                    connections: [{ node: 3, output: 'num', data: {} }],
                },
            },
            outputs: { result: { connections: [] } },
            position: [234, -93],
            name: 'Add',
        },
    },
}
